File: stitch_agent/run/drivers/claude_code.py
"""ClaudeCodeDriver -- delegate fixes to the Claude Code CLI.

Uses --output-format stream-json for real-time streaming of Claude's actions
into the TUI. Each line of stdout is a JSON event; we extract assistant text,
tool use, and tool results to build a human-readable activity log.
"""

from __future__ import annotations

import asyncio
import contextlib
import json
import shutil
from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from stitch_agent.run.drivers.base import build_prompt
from stitch_agent.run.models import FixContext, FixOutcome

if TYPE_CHECKING:
    from collections.abc import Callable


@dataclass
class ClaudeCodeDriver:
    name: str = "claude"
    timeout_seconds: float = 600.0
    binary: str = "claude"
    on_output: Callable[[str], None] | None = field(default=None, repr=False)

    async def fix(self, context: FixContext) -> FixOutcome:
        if not shutil.which(self.binary):
            return FixOutcome(
                applied=False,
                reason=f"{self.binary} CLI not found in PATH",
            )

        prompt = build_prompt(context)
        try:
            proc = await asyncio.create_subprocess_exec(
                self.binary,
                "-p",
                prompt,
                "--permission-mode",
                "acceptEdits",
                "--output-format",
                "stream-json",
                "--verbose",
                cwd=str(context.repo_root),
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.DEVNULL,
            )
        except FileNotFoundError:
            return FixOutcome(
                applied=False,
                reason=f"{self.binary} CLI not found in PATH",
            )
        except Exception as exc:
            return FixOutcome(
                applied=False,
                reason=f"failed to spawn {self.binary}: {exc}",
            )

        return await self._stream_output(proc)

    async def _stream_output(self, proc: asyncio.subprocess.Process) -> FixOutcome:
        activity: list[str] = []
        result_text = ""
        assert proc.stdout is not None

        async def read_events() -> None:
            nonlocal result_text
            while True:
                raw = await proc.stdout.readline()  # type: ignore[union-attr]
                if not raw:
                    break
                line = raw.decode("utf-8", errors="replace").strip()
                if not line:
                    continue

                event = _parse_event(line)
                if event is None:
                    continue

                if event["kind"] == "text":
                    activity.append(event["content"])
                    self._emit(activity)
                elif event["kind"] == "tool_use":
                    activity.append(f"> {event['content']}")
                    self._emit(activity)
                elif event["kind"] == "tool_result":
                    # Show a truncated preview of the result
                    preview = event["content"][:200]
                    if len(event["content"]) > 200:
                        preview += "..."
                    activity.append(f"  {preview}")
                    self._emit(activity)
                elif event["kind"] == "result":
                    result_text = event["content"]

        try:
            await asyncio.wait_for(read_events(), timeout=self.timeout_seconds)
            await proc.wait()
        except TimeoutError:
            with contextlib.suppress(ProcessLookupError):
                proc.kill()
            return FixOutcome(
                applied=False,
                reason=f"{self.binary} CLI timeout after {self.timeout_seconds}s",
                driver_log="\n".join(activity)[-2000:],
            )

        log = "\n".join(activity)

        if proc.returncode != 0:
            return FixOutcome(
                applied=False,
                reason=f"{self.binary} exited {proc.returncode}",
                driver_log=log[-2000:],
            )

        reason = result_text[:200] if result_text else f"{self.binary} CLI completed"
        return FixOutcome(
            applied=True,
            reason=reason,
            driver_log=log[-2000:],
        )

    def _emit(self, activity: list[str]) -> None:
        if self.on_output:
            self.on_output("\n".join(activity))


def _parse_event(line: str) -> dict[str, str] | None:
    """Extract a human-readable event from a stream-json line."""
    try:
        data = json.loads(line)
    except (json.JSONDecodeError, ValueError):
        return None

    event_type = data.get("type")

    if event_type == "assistant":
        msg = data.get("message", {})
        content_blocks = msg.get("content", [])
        parts: list[str] = []
        for block in content_blocks:
            if block.get("type") == "text":
                text = block.get("text", "")
                if text:
                    parts.append(text)
            elif block.get("type") == "tool_use":
                tool_name = block.get("name", "tool")
                tool_input = block.get("input", {})
                desc = tool_input.get("description") or tool_input.get("command") or ""
                if desc:
                    return {"kind": "tool_use", "content": f"{tool_name}: {desc}"}
                # For tools like Edit, show the file path
                path = tool_input.get("file_path") or tool_input.get("path") or ""
                pattern = tool_input.get("pattern") or tool_input.get("query") or ""
                if path:
                    return {"kind": "tool_use", "content": f"{tool_name}: {path}"}
                if pattern:
                    return {"kind": "tool_use", "content": f"{tool_name}: {pattern}"}
                return {"kind": "tool_use", "content": tool_name}
        if parts:
            return {"kind": "text", "content": " ".join(parts)}
        return None

    if event_type == "user":
        # Tool results come back as user messages
        msg = data.get("message", {})
        content_blocks = msg.get("content", [])
        for block in content_blocks:
            if block.get("type") == "tool_result":
                result_content = block.get("content", "")
                if isinstance(result_content, list):
                    result_content = " ".join(
                        b.get("text", "") for b in result_content if isinstance(b, dict)
                    )
                if isinstance(result_content, str) and result_content.strip():
                    return {"kind": "tool_result", "content": result_content.strip()}
        return None

    if event_type == "result":
        return {"kind": "result", "content": data.get("result", "")}

    return None

File: src/drivers/claude-code.ts
import { spawn } from "node:child_process";
import type { FixContext, FixOutcome } from "../core/models.js";
import { which } from "../util.js";
import { buildPrompt } from "./prompt.js";
import type { AgentDriver } from "./types.js";

export class ClaudeCodeDriver implements AgentDriver {
  name = "claude";
  timeoutSeconds: number;
  binary: string;
  onOutput: ((log: string) => void) | null = null;

  constructor(binary = "claude", timeoutSeconds = 600) {
    this.binary = binary;
    this.timeoutSeconds = timeoutSeconds;
  }

  async fix(context: FixContext, signal?: AbortSignal): Promise<FixOutcome> {
    if (!which(this.binary)) {
      return { applied: false, reason: `${this.binary} CLI not found in PATH`, driverLog: "" };
    }
    if (signal?.aborted) {
      return { applied: false, reason: `${this.binary} aborted before start`, driverLog: "" };
    }

    const prompt = buildPrompt(context);
    return new Promise((resolve) => {
      const proc = spawn(
        this.binary,
        [
          "-p",
          prompt,
          "--permission-mode",
          "acceptEdits",
          "--output-format",
          "stream-json",
          "--verbose",
        ],
        { cwd: context.repoRoot, stdio: ["ignore", "pipe", "pipe"] },
      );

      const activity: string[] = [];
      let resultText = "";
      let done = false;

      const onAbort = () => {
        if (done) return;
        done = true;
        clearTimeout(timer);
        signal?.removeEventListener("abort", onAbort);
        try {
          proc.kill("SIGKILL");
        } catch {
          // ignore
        }
        resolve({
          applied: false,
          reason: `${this.binary} aborted`,
          driverLog: activity.join("\n").slice(-2000),
        });
      };
      signal?.addEventListener("abort", onAbort, { once: true });

      const timer = setTimeout(() => {
        if (!done) {
          done = true;
          signal?.removeEventListener("abort", onAbort);
          try {
            proc.kill("SIGKILL");
          } catch {
            // ignore
          }
          resolve({
            applied: false,
            reason: `${this.binary} CLI timeout after ${this.timeoutSeconds}s`,
            driverLog: activity.join("\n").slice(-2000),
          });
        }
      }, this.timeoutSeconds * 1000);

      let buffer = "";
      proc.stdout?.on("data", (chunk: Buffer) => {
        buffer += chunk.toString("utf-8");
        const lines = buffer.split("\n");
        buffer = lines.pop() ?? "";
        for (const line of lines) {
          resultText = handleStdoutLine(line, activity, resultText, (a) => this.emit(a));
        }
      });

      proc.on("close", (code) => {
        if (done) return;
        done = true;
        clearTimeout(timer);
        signal?.removeEventListener("abort", onAbort);
        const log = activity.join("\n");

        if (code !== 0) {
          resolve({
            applied: false,
            reason: `${this.binary} exited ${code}`,
            driverLog: log.slice(-2000),
          });
          return;
        }

        resolve({
          applied: true,
          reason: resultText.slice(0, 200) || `${this.binary} CLI completed`,
          driverLog: log.slice(-2000),
        });
      });

      proc.on("error", (err) => {
        if (done) return;
        done = true;
        clearTimeout(timer);
        signal?.removeEventListener("abort", onAbort);
        resolve({
          applied: false,
          reason: `failed to spawn ${this.binary}: ${err}`,
          driverLog: "",
        });
      });
    });
  }

  private emit(activity: string[]): void {
    if (this.onOutput) {
      this.onOutput(activity.join("\n"));
    }
  }
}

export interface ParsedEvent {
  kind: "text" | "tool_use" | "tool_result" | "result";
  content: string;
}

export function handleStdoutLine(
  line: string,
  activity: string[],
  resultText: string,
  emit: (activity: string[]) => void,
): string {
  if (!line.trim()) return resultText;
  const event = parseEvent(line.trim());
  if (!event) return resultText;
  if (event.kind === "text") {
    activity.push(event.content);
    emit(activity);
  } else if (event.kind === "tool_use") {
    activity.push(`> ${event.content}`);
    emit(activity);
  } else if (event.kind === "tool_result") {
    const preview =
      event.content.length > 200 ? `${event.content.slice(0, 200)}...` : event.content;
    activity.push(`  ${preview}`);
    emit(activity);
  } else if (event.kind === "result") {
    return event.content;
  }
  return resultText;
}

function parseToolUseBlock(b: Record<string, unknown>): ParsedEvent {
  const toolName = (b.name as string) ?? "tool";
  const input = (b.input as Record<string, unknown>) ?? {};
  const label =
    (input.description as string) ||
    (input.command as string) ||
    (input.file_path as string) ||
    (input.path as string) ||
    (input.pattern as string) ||
    (input.query as string) ||
    "";
  return { kind: "tool_use", content: label ? `${toolName}: ${label}` : toolName };
}

function parseAssistantEvent(data: Record<string, unknown>): ParsedEvent | null {
  const msg = data.message as Record<string, unknown> | undefined;
  const contentBlocks = (msg?.content as unknown[]) ?? [];
  const parts: string[] = [];

  for (const block of contentBlocks) {
    if (typeof block !== "object" || block === null) continue;
    const b = block as Record<string, unknown>;
    if (b.type === "text" && typeof b.text === "string" && b.text) {
      parts.push(b.text);
    } else if (b.type === "tool_use") {
      return parseToolUseBlock(b);
    }
  }
  if (parts.length > 0) return { kind: "text", content: parts.join(" ") };
  return null;
}

function parseUserEvent(data: Record<string, unknown>): ParsedEvent | null {
  const msg = data.message as Record<string, unknown> | undefined;
  const contentBlocks = (msg?.content as unknown[]) ?? [];
  for (const block of contentBlocks) {
    if (typeof block !== "object" || block === null) continue;
    const b = block as Record<string, unknown>;
    if (b.type === "tool_result") {
      let resultContent = b.content;
      if (Array.isArray(resultContent)) {
        resultContent = resultContent
          .filter((x): x is Record<string, unknown> => typeof x === "object" && x !== null)
          .map((x) => (x.text as string) ?? "")
          .join(" ");
      }
      if (typeof resultContent === "string" && resultContent.trim()) {
        return { kind: "tool_result", content: resultContent.trim() };
      }
    }
  }
  return null;
}

export function parseEvent(line: string): ParsedEvent | null {
  let data: Record<string, unknown>;
  try {
    data = JSON.parse(line);
  } catch {
    return null;
  }

  const eventType = data.type;

  if (eventType === "assistant") return parseAssistantEvent(data);
  if (eventType === "user") return parseUserEvent(data);
  if (eventType === "result") return { kind: "result", content: (data.result as string) ?? "" };

  return null;
}
